TooltipDecorator/useTooltip.js

```diff
@@ -0,0 +1,266 @@
+	delete props.tooltipProps;
+	delete props.tooltipRelative;
+	delete props.tooltipText;
+	delete props.tooltipType;
+	delete props.tooltipUpdateDelay;
+	delete props.tooltipWidth;
+
+	return props;
+};
+
+// A hook to show Sandstone-styled tooltip components.
+function useTooltip (props = {}) {
+	const {
+		screenEdgeKeepout = (24 + 24), // Do NOT forget to check TooltipDecorator's default config value also.
+		tooltipDelay = 500, tooltipType = 'balloon', tooltipUpdateDelay = 400,
+		tooltipMarquee, tooltipPosition, tooltipProps, tooltipRelative, tooltipText, tooltipWidth
+	} = props;
+	const rtl = useI18nContext()?.rtl;
+
+	const [showing, setShowing] = useState(false);
+	const [layoutInfo, setLayoutInfo] = useState({
+		arrowAnchor: null,
+		labelOffset: null,
+		position: {top: 0, left: 0},
+		tooltipDirection: null
+	});
+	const mutableRef = useRef({
+		mutationObserver: null,
+		resizeObserver: null,
+		showTooltipJob: null,
+		setTooltipLayoutJob: null
+	});
+	const clientRef = useRef(null);
+	const tooltipRef = useRef(null);
+
+	const showTooltip = useCallback((client) => {
+		if (tooltipText) {
+			clientRef.current = client;
+			currentTooltip = client;
+			mutableRef.current.showTooltipJob?.startAfter(tooltipDelay);
+			mutableRef.current.mutationObserver?.observe(clientRef.current, {attributes: true, childList: true});
+			mutableRef.current.resizeObserver?.observe(clientRef.current);
+		}
+	}, [clientRef, tooltipDelay, tooltipText]);
+
+	const hideTooltip = useCallback(() => {
+		if (tooltipText) {
+			mutableRef.current.mutationObserver?.disconnect();
+			mutableRef.current.resizeObserver?.disconnect();
+
+			clientRef.current = null;
+			currentTooltip = null;
+
+			mutableRef.current.showTooltipJob?.stop();
+			mutableRef.current.setTooltipLayoutJob?.stop();
+
+			setShowing(false);
+		}
+	}, [clientRef, tooltipText]);
+
+	const startTooltipLayoutJob = useCallback(() => {
+		mutableRef.current.setTooltipLayoutJob?.startAfter(tooltipUpdateDelay);
+	}, [tooltipUpdateDelay]);
+
+	const handlers = useHandlers({
+		// Recalculate tooltip layout on keydown to make sure tooltip is positioned correctly in case something changes as a result of the keydown.
+		onKeyDown: handle(
+			forward('onKeyDown'),
+			not(forProp('disabled', true)),
+			() => {
+				startTooltipLayoutJob();
+			}
+		),
+		onMouseOver: handle(
+			forward('onMouseOver'),
+			forProp('disabled', true),
+			(ev) => {
+				showTooltip(ev.currentTarget);
+			}
+		),
+		onMouseOut: handle(
+			forward('onMouseOut'),
+			forProp('disabled', true),
+			(ev) => {
+				if (clientRef.current && !clientRef.current.contains(ev.relatedTarget)) {
+					hideTooltip();
+				}
+			}
+		),
+		onFocus: handle(
+			forward('onFocus'),
+			({target}) => showTooltip(target)
+		),
+		onBlur: handle(
+			forward('onBlur'),
+			hideTooltip
+		)
+	}, props);
+
+	useEffect(() => {
+		const mutableRefCurrent = mutableRef.current;
+
+		if (window.MutationObserver) {
```

Comment:
Please check window existence with `typeof`.

Reply:
Good catch! I missed this. Thank you!